neo/io/nwbio.py

```diff
@@ -0,0 +1,823 @@
+        else:
+            timeseries_class = TimeSeries  # default
+        additional_metadata = {name[4:]: value
+                               for name, value in signal.annotations.items()
+                               if name.startswith("nwb:")}
+        if "nwb_electrode" in signal.annotations:
+            electrode_name = signal.annotations["nwb_electrode"]["name"]
+            additional_metadata["electrode"] = electrodes[electrode_name]
+        if timeseries_class != TimeSeries:
+            conversion, units = get_units_conversion(signal, timeseries_class)
+            additional_metadata["conversion"] = conversion
+        else:
+            units = signal.units
+        if isinstance(signal, AnalogSignal):
+            sampling_rate = signal.sampling_rate.rescale("Hz")
+            tS = timeseries_class(
+                name=signal.name,
+                starting_time=time_in_seconds(signal.t_start),
+                data=signal,
+                unit=units.dimensionality.string,
+                rate=float(sampling_rate),
+                comments=json.dumps(hierarchy),
+                **additional_metadata)
+                # todo: try to add array_annotations via "control" attribute
+        elif isinstance(signal, IrregularlySampledSignal):
+            tS = timeseries_class(
+                name=signal.name,
+                data=signal,
+                unit=units.dimensionality.string,
+                timestamps=signal.times.rescale('second').magnitude,
+                comments=json.dumps(hierarchy),
+                **additional_metadata)
+        else:
+            raise TypeError("signal has type {0}, should be AnalogSignal or IrregularlySampledSignal".format(
+                signal.__class__.__name__))
+        nwb_group = signal.annotations.get("nwb_group", "acquisition")
+        add_method_map = {
+            "acquisition": nwbfile.add_acquisition,
+            "stimulus": nwbfile.add_stimulus
+        }
+        if nwb_group in add_method_map:
+            add_time_series = add_method_map[nwb_group]
+        else:
+            raise NotImplementedError("NWB group '{}' not yet supported".format(nwb_group))
+        add_time_series(tS)
+        return tS
+
+    def _write_spiketrain(self, nwbfile, spiketrain):
+        nwbfile.add_unit(spike_times=spiketrain.rescale('s').magnitude,
+                         obs_intervals=[[float(spiketrain.t_start.rescale('s')),
+                                         float(spiketrain.t_stop.rescale('s'))]],
+                         _name=spiketrain.name,
+                         # _description=spiketrain.description,
+                         segment=spiketrain.segment.name,
+                         block=spiketrain.segment.block.name)
+        # todo: handle annotations (using add_unit_column()?)
+        # todo: handle Neo Units
+        # todo: handle spike waveforms, if any (see SpikeEventSeries)
```

Comment:
You should be able to just add a `waveforms` field here.

Reply:
https://pynwb.readthedocs.io/en/stable/pynwb.misc.html?highlight=Units#pynwb.misc.Units.add_unit